File: skills/openclix-analytics/assets/flutter/lib/openclix/analytics/openclix_analytics_emitter.dart
typedef OpenClixAnalyticsSink = Future<void> Function(
  String eventName,
  Map<String, Object?> properties,
);

enum OpenClixSourceType { app, system }
enum OpenClixAnalysisPeriod { pre, post }

class OpenClixAnalyticsEvent {
  final String name;
  final OpenClixSourceType sourceType;
  final Map<String, Object?> properties;

  const OpenClixAnalyticsEvent({
    required this.name,
    required this.sourceType,
    this.properties = const {},
  });
}

class OpenClixAnalyticsEmitter {
  final String platform;
  final OpenClixAnalysisPeriod analysisPeriod;
  final bool campaignActive;
  final OpenClixAnalyticsSink sink;
  final String Function(String canonicalName)? eventNameTransform;

  const OpenClixAnalyticsEmitter({
    required this.platform,
    required this.analysisPeriod,
    required this.campaignActive,
    required this.sink,
    this.eventNameTransform,
  });

  Future<void> emit(OpenClixAnalyticsEvent event) async {
    final merged = <String, Object?>{
      ...event.properties,
      'openclix_source': 'openclix',
      'openclix_event_name': event.name,
      'openclix_source_type': event.sourceType.name,
      'openclix_platform': platform,
      'openclix_campaign_id': _stringOrNull(event.properties['campaign_id']),
      'openclix_queued_message_id':
          _stringOrNull(event.properties['queued_message_id']),
      'openclix_channel_type': _stringOrNull(event.properties['channel_type']),
      'openclix_analysis_period': analysisPeriod.name,
      'openclix_campaign_active': campaignActive ? 'true' : 'false',
    };

    final outboundName = eventNameTransform?.call(event.name) ?? event.name;
    await sink(outboundName, merged);
  }

  String? _stringOrNull(Object? value) {
    return value is String ? value : null;
  }
}

String normalizeFirebaseEventName(String name) {
  var normalized = name.toLowerCase().replaceAll(RegExp(r'[^a-z0-9_]'), '_');
  if (!RegExp(r'^[a-z]').hasMatch(normalized)) {
    normalized = 'oc_$normalized';
  }
  if (normalized.length > 40) {
    normalized = normalized.substring(0, 40);
  }
  return normalized;
}

File: skills/openclix-analytics/assets/android/app/src/main/kotlin/ai/openclix/analytics/OpenClixAnalyticsEmitter.kt
package ai.openclix.analytics

enum class OpenClixSourceType {
    APP,
    SYSTEM,
}

enum class OpenClixAnalysisPeriod {
    PRE,
    POST,
}

data class OpenClixAnalyticsEvent(
    val name: String,
    val sourceType: OpenClixSourceType,
    val properties: Map<String, Any?> = emptyMap(),
)

class OpenClixAnalyticsEmitter(
    private val platform: String,
    private val analysisPeriod: OpenClixAnalysisPeriod,
    private val campaignActive: Boolean,
    private val sink: (eventName: String, properties: Map<String, Any?>) -> Unit,
    private val eventNameTransform: ((canonicalName: String) -> String)? = null,
) {
    fun emit(event: OpenClixAnalyticsEvent) {
        val merged = event.properties.toMutableMap()
        merged["openclix_source"] = "openclix"
        merged["openclix_event_name"] = event.name
        merged["openclix_source_type"] = event.sourceType.name.lowercase()
        merged["openclix_platform"] = platform
        merged["openclix_campaign_id"] = event.properties["campaign_id"] as? String
        merged["openclix_queued_message_id"] = event.properties["queued_message_id"] as? String
        merged["openclix_channel_type"] = event.properties["channel_type"] as? String
        merged["openclix_analysis_period"] = analysisPeriod.name.lowercase()
        merged["openclix_campaign_active"] = if (campaignActive) "true" else "false"

        val outboundName = eventNameTransform?.invoke(event.name) ?: event.name
        sink(outboundName, merged)
    }
}

fun normalizeFirebaseEventName(name: String): String {
    val normalized = buildString {
        for (ch in name.lowercase()) {
            append(if (ch.isLetterOrDigit() || ch == '_') ch else '_')
        }
    }

    val withPrefix = if (normalized.firstOrNull()?.isLetter() == true) {
        normalized
    } else {
        "oc_$normalized"
    }

    return withPrefix.take(40)
}
